### PR TITLE
Adding torch backend decorator to handle numpy arrays

### PR DESCRIPTION
**Pull request by abdulasiraj**

### CODE REVIEW

**Comment by KareemMAX** [comment]
ivy-gardener

**Comment by KareemMAX** [comment]
Hi @abdulasiraj, I was testing linting on your PR assuming I could make it green 😅

Please ignore the lint-bot, and the changes it made.

**Comment by abdulasiraj** [comment]
Hi @CatB1t ,
can you please make some comments on this?
Thanks!

**Comment by abdulasiraj** [comment]
Hi @CatB1t ,
Can you please have a look?

File: ivy/func_wrapper.py [hunk 1]
```diff
@@ -8,6 +8,11 @@
 from types import FunctionType
 from typing import Callable
 import inspect
+import torch
```
Comment (CatB1t) [inline ivy/func_wrapper.py:11]: We shouldn't import torch here, let's make it backend specific decorator

File: ivy/func_wrapper.py [hunk 2]
```diff
@@ -660,6 +665,34 @@ def _handle_view_indexing(*args, **kwargs):
     return _handle_view_indexing
 
 
+def _convert_numpy_arrays_to_torch(*args):
+    torch_tensors = []
+    for array in args:
+        if isinstance(array, np.ndarray):
```
Comment (CatB1t) [inline ivy/func_wrapper.py:671]: Ideally, we should do a nested search as the argument may be a list containing numpy arrays, which wouldn't work in that case. maybe make use of ivy nest functions here.

**Review by CatB1t — CHANGES_REQUESTED** [review]

File: ivy/func_wrapper.py (hunk 2 above)
Comment (VedPatwardhan) [reply ivy/func_wrapper.py:671]: Adding to this, it would be great if we could try and reuse/refactor some of the logic [here](https://github.com/unifyai/ivy/blob/d5cc0291e64af78f58daef7650d9f75a6625df50/ivy/utils/backend/handler.py#L366) at it serves a similar purpose (except for the garbage collection part). Thanks!

**Review by VedPatwardhan — CHANGES_REQUESTED** [review]

**Comment by abdulasiraj** [comment]
Hi @CatB1t @VedPatwardhan ,

It's working very fine now and ready for review. suggestions were really helpful. Thanks all!

File: ivy/func_wrapper.py (hunk 2 above)
Comment (VedPatwardhan) [reply ivy/func_wrapper.py:667]: I think what @CatB1t meant here is that we do something like,
```
np_arr_idxs = ivy.nested_argwhere(args)        # nested search
np_arr_vals = ivy.multi_index_nest(args)         # indexing to get the arrays  
backend_arr_vals = [ivy.asarray(val).to_native() for val in np_arr_vals]      # conversion
ivy.set_nest_at_indices(args, np_arr_idxs, backend_arr_vals)   # setting the backend ones back
```

**Review by VedPatwardhan — CHANGES_REQUESTED** [review]
Hey @abdulasiraj , just requsted a change, rest looks good to go. Thanks 🙂

**Comment by abdulasiraj** [comment]
Hi @VedPatwardhan 
This one is ready for review.
Thanks!

File: ivy/functional/backends/torch/elementwise.py [hunk 3]
```diff
@@ -48,7 +50,8 @@ def bitwise_xor(
 bitwise_xor.support_native_out = True
 
 
-@with_unsupported_dtypes({"2.0.1 and below": ("float16", "complex")}, backend_version)
```
Comment (VedPatwardhan) [inline ivy/functional/backends/torch/elementwise.py:51]: Could you please revert the changes made to `with_unsupported_dtypes` in this file?

**Review by VedPatwardhan — CHANGES_REQUESTED** [review]
Hey @abdulasiraj, just a small change, could you please also look into why all the intelligent tests are failing and whether the failures are related to the changes you've made? I'll also add the `Exhaustive CI` label just to be sure that it doesn't break anything. Thanks 🙂

**Comment by abdulasiraj** [comment]
Hi @VedPatwardhan 
It's ready to review.

**Comment by VedPatwardhan** [comment]
> Hi @VedPatwardhan It's ready to review.

Hey @abdulasiraj, could you please link the task that involves fixing this? Because given that this would add quite a bit of overhead to `torch` functions, it would be useful to have this as an optional mode because most usages wouldn't necessarily involve `numpy` arrays. Thanks🙂

**Comment by abdulasiraj** [comment]
> > Hi @VedPatwardhan It's ready to review.
> 
> Hey @abdulasiraj, could you please link the task that involves fixing this? Because given that this would add quite a bit of overhead to `torch` functions, it would be useful to have this as an optional mode because most usages wouldn't necessarily involve `numpy` arrays. Thanks 🙂

https://trello.com/c/1JPoKxID

**Review by VedPatwardhan — COMMENTED** [review]
Sounds good! Is this limited to elementwise functions? If not we might want to add it for all such functions which work with `np.ndarray` for the other backends. Thanks @abdulasiraj 🙂
